python/ray/tests/test_generators.py

```diff
@@ -80,20 +80,144 @@ def generator(num_returns):
+        assert ray.get(fetch.remote(ref)) == i
+
+    cluster.add_node(num_cpus=1, resources={"node2": 1}, object_store_memory=10 ** 8)
+
+    # Fetch one of the ObjectRefs to another node. We should try to reuse this
+    # copy during recovery.
+    ray.get(fetch.options(resources={"node2": 1}).remote(refs[-1]))
+    cluster.remove_node(node_to_kill, allow_graceful=False)
+    for i, ref in enumerate(refs):
+        assert ray.get(fetch.remote(ref)) == i
+
+
+@pytest.mark.parametrize("too_many_returns", [False, True])
+def test_dynamic_generator_reconstruction_nondeterministic(
+    ray_start_cluster, too_many_returns
+):
+    config = {
+        "num_heartbeats_timeout": 10,
+        "raylet_heartbeat_period_milliseconds": 100,
+        "max_direct_call_object_size": 100,
+        "task_retry_delay_ms": 100,
+        "object_timeout_milliseconds": 200,
+        "fetch_warn_timeout_milliseconds": 1000,
+    }
+    cluster = ray_start_cluster
+    # Head node with no resources.
+    cluster.add_node(
+        num_cpus=0,
+        _system_config=config,
+        enable_object_reconstruction=True,
+        resources={"head": 1},
+    )
+    ray.init(address=cluster.address)
+    # Node to place the initial object.
+    node_to_kill = cluster.add_node(num_cpus=1, object_store_memory=10 ** 8)
+    cluster.wait_for_nodes()
+
+    @ray.remote(num_cpus=0, resources={"head": 1})
+    class ExecutionCounter:
+        def __init__(self):
+            self.count = 0
+
+        def inc(self):
+            self.count += 1
+            return self.count
+
+    @ray.remote(num_returns="dynamic")
+    def dynamic_generator(exec_counter):
+        num_returns = 10
+        if ray.get(exec_counter.inc.remote()) > 1:
+            if too_many_returns:
+                num_returns += 1
+            else:
+                num_returns -= 1
+        for i in range(num_returns):
+            yield np.ones(1_000_000, dtype=np.int8) * i
+
+    @ray.remote
+    def fetch(x):
+        return
+
+    exec_counter = ExecutionCounter.remote()
+    gen = dynamic_generator.remote(exec_counter)
+    cluster.remove_node(node_to_kill, allow_graceful=False)
+    node_to_kill = cluster.add_node(num_cpus=1, object_store_memory=10 ** 8)
+    refs = list(gen)
+    if too_many_returns:
+        for ref in refs:
+            ray.get(ref)
+    else:
+        with pytest.raises(ray.exceptions.RayTaskError):
+            for ref in refs:
+                ray.get(ref)
+    # TODO(swang): If the re-executed task returns a different number of
+    # objects, we should throw an error for every return value.
+    # for ref in refs:
+    #     with pytest.raises(ray.exceptions.RayTaskError):
+    #         ray.get(ref)
+
+
+# TODO: Test cases:
+# - test exception when num_returns="dynamic" but not a generator, and vice versa
+# - generator errors before yield
+# - generator calls ray.put, reconstruction
+# - ref counting, check for leaks
+# - passing ObjRefGenerator, passing generated ObjRefs
```

Comment:
also wondering the behavior when generator failed after yield a few objects. would retry ever work in that case?

Reply:
Yeah right now it won't work if the earlier objects were stored in plasma. Added a test case and TODO to show this behavior.